diag/diag.proto

```diff
@@ -0,0 +1,238 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// This file defines the gNOI APIs used to perform diagnostic operations on a
+// switching box.
+syntax = "proto3";
+
+package gnoi.diag;
+
+// The Diag service exports to main set of RPCs:
+// 1- BERT related RPCs: Used to perform Bit Error Rate Test (BERT)
+//    operations on a set of ports.
+// 2- BURNING related RPCs: Used to perform a vendor-provided Burnin test on the
+//    switch to ensure the box is ready to start serving traffic. Burnin tests
+//    are typically run in the datacenter, as part of turnup or repair workflow.
+// Note: The RPCs defined here are are stateless operations and them
+// failing/passing should not leave any permanent artifact on the switch (unless
+// there is something wrong HW-wise).
+service Diag {
+  // Starts BERT operation on a set of ports. Each BERT operation is uniquely
+  // identified by an ID, which is given by the caller. The caller can then
+  // use this ID (as well as the list of the ports) to stop the BERT operation
+  // and/or get the BERT results. This RPC is expected to return an error status
+  // in the following situations:
+  // - When BERT operation is not supported on any of the ports specified by
+  //   the request.
+  // - When BERT is already in progress on any port specified by the request.
+  // - In case of any low-level HW/SW internal errors.
+  // The RPC returns an OK status of none of these situations is encountered.
+  rpc StartBERT(StartBERTRequest) returns(StartBERTResponse) {}
+
+  // Stops an already in-progress BERT operation on a set of ports. The caller
+  // uses the BERT operation ID it previously used when starting the operation
+  // to stop it. The RPC is expected to return an error status in the following
+  // situations:
+  // - When there is at least one BERT operation in progress on a port which
+  //   cannot be stopped in the middle of the operation (either due to lack of
+  //   support or internal problems).
+  // - When no BERT operation, which matches the given BERT operation ID, is in
+  //   progress or completed on any of the ports specified by the request.
+  // - When the BERT operation ID does not match the in progress or completed
+  //   BERT operation on any of the ports specified by the request.
+  // The RPC returns an OK status of none of these situations is encountered.
+  // Note that a BERT operation is considered completed if switch has a
+  // record of it. Also note that it is OK to receive a stop request for a port
```

Comment:
I'm not clear what "if switch has a record of it" means. Does this mean to say that only BERT operations that the switch has the history of are considered completed. For example, a BERT operation that ran before the last reboot might be unknown?

Reply:
Yes you got it right. After a BERT operation is done, there is nothing in HW that keeps track of the completed BERTs. So it is up to the agent to keep track of them. The agent typically has a limited resource and keeps track of the last M BERTs. Also the storage used to keep track of BERTs might be volatile (wiped out after reboot). At any point of time if the switch knows about a previous BERT, it can consider this completed. Otherwise it is considered a new BERT. 

I added more details on this. PTAL.